deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,647 @@
+			prev_it->ai_next = it->ai_next;
+			it->ai_next = cur;
+		}
+		prev = cur;
+		cur = cur->ai_next;
+	}
+
+	return STATUS_SUCCESS;
+}
+
+/**
+ *	Takes the errors that may have been generated 
+ *
+ */
+static int coalesce_errors(struct happy_eyeballs_ctx *context)
+{
+	// Don't coalesce errors while starting
+	if (os_atomic_load_bool(&context->is_starting))
+		return STATUS_FAILURE;
+
+	// We'll use the mode of the errors for now.
+	struct mode {
+		int error;
+		int count;
+	};
+	DARRAY(struct mode) modes = {0};
+	da_init(modes);
+	// Gather all the errors into counts for each error
+	pthread_mutex_lock(&context->candidate_mutex);
+	for (int i = 0; i < context->candidates.num; i++) {
+		int err = context->candidates.array[i].error;
+		if (err == 0)
+			// Error is 0, just move on.
+			continue;
+
+		struct mode *mode = NULL;
+		// Find an existing index containing this error, if it exists
+		for (int j = 0; j < modes.num && mode == NULL; j++) {
+			if (modes.array[j].error == err)
+				mode = &modes.array[j];
+		}
+		if (mode == NULL)
+			// Existing index doesn't exist, take the next available.
+			mode = da_push_back_new(modes);
+
+		// Note the error code and increment the count.
+		mode->error = err;
+		mode->count++;
+	}
+	pthread_mutex_unlock(&context->candidate_mutex);
+	int max_count = 0;
+	int max_value = 0;
+	// Find the error with the most occurrences.
+	for (int i = 0; i < modes.num; i++) {
+		if (max_count < modes.array[i].count) {
+			max_value = modes.array[i].error;
+			max_count = modes.array[i].count;
+		}
+	}
+	da_free(modes);
+	// Set the error
+	context->error = max_value;
+	context->error_message = strerror(context->error);
+
+	return STATUS_SUCCESS;
+}
+
+static void *happy_connect_worker(void *arg)
+{
+	struct happy_connect_woker_args *args =
+		(struct happy_connect_woker_args *)arg;
+	struct happy_eyeballs_ctx *context = args->context;
+	if (args->sockfd == INVALID_SOCKET) {
+		free(args);
+		return NULL;
+	}
+	if (os_event_try(args->context->race_completed_event) == 0) {
+		// Already lost, don't bother
+		free(args);
+		return NULL;
+	}
+
+#if defined(_WIN32) && defined(SO_NOSIGPIPE)
+	setsockopt(args->sockfd, SOL_SOCKET, SO_NOSIGPIPE, &(int){1},
+		   sizeof(int));
+#endif
+	if (context->bind_addr_len &&
+	    bind(args->sockfd, (const struct sockaddr *)&context->bind_addr,
+		 context->bind_addr_len) < 0) {
+		goto failure;
+	}
+	if (connect(args->sockfd, args->address->ai_addr,
+		    (int)args->address->ai_addrlen) == 0) {
+		// success, check if we're the winner.
+		pthread_mutex_lock(&context->winner_mutex);
+		os_event_signal(args->candidate->socket_completed_event);
+		if (os_event_try(context->race_completed_event) == EAGAIN) {
+			// We are the winner.
+			context->socket_fd = args->sockfd;
+			memcpy(&context->winner_addr, args->address->ai_addr,
+			       sizeof(struct addrinfo));
+			os_event_signal(context->race_completed_event);
+		}
+		pthread_mutex_unlock(&context->winner_mutex);
+		free(args);
```

Comment:
maybe this is `goto success` which looks like 
```cpp
success:
    free(args);
    return NULL;
```

and put the label at the bottom of function?